webdriver-spec.html

```diff
@@ -1435,288 +1435,532 @@
 
-   <dt>"<code>acceptSslCerts</code>"
-   <dd>False
-  </dl>
+<p>When <dfn>processing capabilities</dfn> with argument <var>parameters</var>,
+  the <a>endpoint node</a> <em>MUST</em> run the following steps:
 
- <li><p>Let <var>required capabilities</var> be the result of <a>getting a property</a>
-  named <code>requiredCapabilities</code> from <var>capabilities</var>.
+<ol>
+<li><p>Let <var>capabilities request</var> be the result of getting a property
+named "<code>capabilities</code>" from <var>parameters</var>.
 
- <li>If <var>required capabilities</var> is not a JSON Object,
-  set the value to an empty JSON Object.
+  <ol>
+  <li><p>If <var>capabilities request</var> is not a JSON object,
+  return <a>error</a> with <a>error code</a> <a>invalid argument</a>.
+  </ol>
 
- <li><p>Let <var>desired capabilities</var> be the result of <a>getting a property</a>
-  named <code>desiredCapabilities</code> from <var>capabilities</var>.
+<li><p>Let <var>required capabilities</var> be the result of getting a property
+named "<code>alwaysMatch</code>" from <var>capabilities request</var>.
 
- <li>If <var>desired capabilities</var> is not a JSON Object,
+  <ol>
+  <li><p>If <var>required capabilities</var> is <a>null</a> or <a>undefined</a>,
   set the value to an empty JSON Object.
 
- <li><p>Let <var>length</var> be the length of <var>required capabilities</var>.
-
- <li><p>Let <var>k</var> be 0.
+  <li><p>If <var>required capabilities</var> is not a JSON Object,
+  return <a>error</a> with <a>error code</a> <a>invalid argument</a>.
+  </ol>
 
- <li><p>While <var>k</var> &lt; <var>length</var>:
+<li><p>Let <var>desired capabilities</var> be the result of getting a property
+named "<code>firstMatch</code>" from <var>capabilities request</var>.
 
   <ol>
-   <li><p>Let <var>capability</var> be the entry
-    in <var>required capabilities</var> at index <var>k</var>.
+  <li><p>If <var>desired capabilities</var> is <a>null</a> or <a>undefined</a>,
+  set the value to a JSON List with a single entry: an empty JSON Object.
 
-   <li><p>If the name of the <var>capability</var> is among
-    the names of entries in <var>desired capabilities</var>,
-    remove the corresponding entry from <var>desired capabilities</var>.
-
-   <li>Increase <var>k</var> by 1.
+  <li><p>If <var>desired capabilities</var> is not a JSON List,
+  return <a>error</a> with <a>error code</a> <a>invalid argument</a>.
   </ol>
 
- <li><p>Let <var>unmet capabilities</var> be an empty JSON List.
-
- <li><p>Let <var>unprocessed capabilities</var> be a JSON Object
-  that contains all entries from <var>required capabilities</var>
-  and all entries from <var>desired capabilities</var>.
+<li><p>Let <var>length</var> be the length of <var>desired capabilities</var>.
 
- <li><p>Let <var>j</var> be 0.
+<li><p>Let <var>k</var> be 0.
 
- <li><p>Let <var>capabilities length</var> be
-  the length of <var>unprocessed capabilities</var>.
+<li><p>Let <var>matched capabilities</var> be <a>null</a>.
```

Comment:
The way this is written, if both `firstMatch` and `alwaysMatch` are empty, then `matched capabilities` will by `null` when exiting the loop, meaning an weird value will be returned. ATM, selenium will return a default driver implementation (probably firefox) if this is the case. I'd suggest using a default value of an empty JSON Object instead of `null` in this section.